File: assets/mods/std/section/light2d.mjs
import Color from "package:color/index.js";
import { create as transformCreate } from "std:bevy/transform/transform.mjs";
import { Synchronize } from "std:synchronize.mjs";

export class PointLight2d extends Synchronize {
  constructor(
    transform = transformCreate(),
    radius = 0.5,
    color = Color({ r: 255, g: 255, b: 255, alpha: 0.5 }),
    intensity = 1.0,
    falloff = 0.0,
    cast_shadows = false
  ) {
    super();
    this.transform = transform;
    this.color = color;
    this.radius = radius;
    this.intensity = intensity;
    this.falloff = falloff;
    this.cast_shadows = cast_shadows;
  }
}

export function fromValues(
  transform = create(),
  radius = 0.5,
  color = Color({ r: 255, g: 255, b: 255, alpha: 0.5 }),
  intensity = 1.0,
  falloff = 0.0,
  cast_shadows = false
) {
  return new PointLight2d(
    transform,
    radius,
    color,
    intensity,
    falloff,
    cast_shadows
  );
}

File: simple_warfare_server/src/custom/unit/light2d.rs
use bevy::prelude::*;
use boa_engine::value::TryFromJs;

use crate::bevy_ext::try_from_js::*;
use crate::custom::unit::transform::transform::JsTransform;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Component, Serialize, Deserialize, PartialEq, Reflect, TryFromJs)]
pub struct JsPointLight2d {
    pub transform: JsTransform,
    /// The light's color tint.
    #[boa(from_js_with = "color_try_from_js")]
    pub color: Color,
    /// The intensity of the light. The light's attenutation is multiplied by this value.
    /// The higher the intensity, the brighter the light.
    #[boa(from_js_with = "f32_try_from_js")]
    pub intensity: f32,
    /// The radius of the light. Illumination will only occur within the light's radius.
    #[boa(from_js_with = "f32_try_from_js")]
    pub radius: f32,
    /// How quickly illumination from the light should deteriorate over distance.
    /// A higher falloff value will result in less illumination at the light's maximum radius.
    #[boa(from_js_with = "f32_try_from_js")]
    pub falloff: f32,
    /// Whether the light should cast shadows.
    pub cast_shadows: bool,
}

impl Default for JsPointLight2d {
    fn default() -> Self {
        Self {
            transform: JsTransform::default(),
            color: Color::WHITE,
            intensity: 1.0,
            radius: 0.5,
            falloff: 0.0,
            cast_shadows: false,
        }
    }
}
